README.rst
MICtools
========

.. image:: https://travis-ci.org/minepy/mictools.svg?branch=master
    :target: https://travis-ci.org/minepy/mictools

.. image:: https://badge.fury.io/py/mictools.svg
    :target: https://badge.fury.io/py/mictools

MICtools is an open source pipeline which combines the TIC_e and MIC_e measures
[Reshef2016]_ into a two-step procedure that allows to identify relationships of
various degrees of complexity in large datasets. TIC_e is used to perform 
an efficient high throughput screening of all the possible pairwise
relationships and a permutation based appraoch is used to assess their significance.  
MIC_e is then used to rank the subset of significant associations on the bases of their strength.

**Please cite**: Davide Albanese, Samantha Riccadonna, Claudio Donati, Pietro
Franceschi; A practical tool for Maximal Information Coefficient analysis, 
GigaScience, giy032, https://doi.org/10.1093/gigascience/giy032

.. image:: docs/images/schema.png

The MICtools pipeline can be broken into 4 steps (see the figure above):

#. given M variables pairs x_i and y_i measured in n samples, the empirical
   TIC_e null distribution is estimated by permutation;
#. TIC_e statistics and the associated empirical p-values are computed for all 
   variable pairs;
#. p-values are corrected for multiplicity in order to control the family-wise
   error rate (FWER) or the false discovery rate (FDR);
#. finally, the strengths of the relationships called significant are estimated 
   using the MIC_e estimator.

**Table of contents**

.. contents:: :local:

Install
-------

Using pip (Linux and macOS/OS X)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

We suggest to install Python 3 (3.5+) and the GCC compiler through the package manager 
(In Mac OS X, we recommend to install them using `Homebrew <http://brew.sh/>`_)
(e.g. on Ubuntu/Debian):

.. code-block:: sh

    sudo apt-get update
    sudo apt-get install build-essential python3-dev

Then, upgrade pip and install setuptools:

.. code-block:: sh

    pip install --upgrade pip
    pip install 'setuptools >=14.0'

Finally, install mictools:

.. code-block:: sh

    pip install mictools

Docker (Linux, macOS/OS X and MS Windows)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

#. Install Docker for `Linux <https://docs.docker.com/linux/>`_,
   `Mac OS X <https://docs.docker.com/mac/>`_ or
   `Windows <https://docs.docker.com/windows/>`_.

#. Run the ``Docker Quickstart Terminal`` (Mac OS X, Windows) or the
   ``docker`` daemon (Linux, ``sudo service docker start``).

#. Follow the instructions at https://hub.docker.com/r/minepy/mictools/.

From source
^^^^^^^^^^^

If you are installing from source, the following dependences must be installed:
Python >= 3.5, Click >= 5.1, numpy >= 1.7.0, scipy >= 0.13, pandas >= 0.17.0,
matplotlib >= 1.2.0,<2, statsmodels >= 0.6.1, minepy >= 1.2. We suggest to
install these dependences using the OS package manager (Linux), Homebrew 
(macOS/OS X) or pip.

Download the latest stable version from https://github.com/minepy/mictools/releases
and complete the installation:

.. code-block:: sh

   tar -zxvf mictools-X.Y.Z.tar.gz
   python3 setup.py install

Usage
-----

MICtools can be used to investigate variable associations in different types of experimental scenarios:

* single dataset X, with M variables and N samples: to evaluate
  the M+(M-1)/2 possible associations;
* two datasets, X (MxN) and Y (KxN) (parameter -y/--yvars): to evaluate 
  all the pairwise relationships between the variables of the two
  datasets (for a total of MxK associations). Note that the number samples (N)
  in the datasets X and Y must be the same.
* two datasets, X (MxN) and Y (KxN): to evaluate all the rowwise 
  relationships (see -r/--rowwise), i.e. only the variables pairs X_i and Y_i
  (for each i in min(M, K)) will be tested;
In all the abovementioned cases the analysis will be performed within each class independently if the sample classes are provided (see -l/--labels and -t/--target).

MICtools is implemented as a single command (``mictools'') with the following
subcommands:

``null``
  Compute the TIC_e null distribution.

``mergenull``
  Merge multiple TIC_e null distributions.

``pval``
  Compute TIC_e p-values.

``adjust``
  Multiple testing correction.

``strength``
  Compute the strength (MIC_e).

Run ``mictools SUBCOMMAND --help`` for the documentation of each specific step.

Tutorial
--------
We analyze the "Datasaurus" synthetic dataset generated following the approach discussed at  https://www.autodeskresearch.com/publications/samestats
([Matejka2017]_). The dataset contains 26 variables linked by 13 relationships
which have the same summary statistics (e.g. the Pearson's correlation),
but are very different in appearance. The dataset was modified in order to 
destroy secondary associations. In this example we test the entire set of possible 
associations (for a total of 26*(26-1)/2 = 325 relationships).

Preparation 
^^^^^^^^^^^
Go to the ``examples`` folder:

.. code-block:: sh

  cd examples

Select the Datasaurus dataset and the output folder:

.. code-block:: sh

  X=datasaurus.txt
  ODIR=datasaurus_results
  mkdir $ODIR

Empirical TIC_e null distribution
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Compute the empirical TIC_e null distribution (with 200,000 permutations,
default value):

.. code-block:: sh

  mictools null $X $ODIR/null_dist.txt

The output file ``null_dist.txt`` is a TAB-delimited file which contains the 
null distrubution:

===== ======== ======== ========= ============
Class BinStart BinEnd   NullCount NullCountCum
===== ======== ======== ========= ============
None  0.000000 0.000100 0         200000
None  0.000100 0.000200 0         200000
None  0.000200 0.000300 0         200000
...   ...      ...      ...       ...
===== ======== ======== ========= ============

The first column (``Class``) contains the class membership (in this particular 
case no sample classes were provided), ``BinStart`` and ``BinEnd`` define the
TIC_e range and ``NullCount`` and ``NullCountCum`` are distribution and the 
cumulative distribution, respectively.

TIC_e p-values
^^^^^^^^^^^^^^
Compute the TIC_e statistics and the associated empirical p-values for all 
variable pairs:

.. code-block:: sh

  mictools pval $X $ODIR/null_dist.txt $ODIR

The command will return in the output directory the following:

``obs_dist.txt``
  the observed TICe distribution in the same format of ``null_dist.txt``
  
``obs.txt``
  TAB-delimited file containing the observed TICe values for each variable pair
  tested:

  ====== ========== ========
  Var1   Var2       None
  ====== ========== ========
  away_x bullseye_x	0.029476
  away_x circle_x   0.018211
  away_x dino_x     0.050720
  ...    ...        ...
  ====== ========== ========

``pval.txt``
  TAB-delimited file containing the empirical p-values for each variable pair

``pval_None.png``
  the p values distribution plot:

  .. image:: docs/images/pval_None.png

Multiple testing correction
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Correct the p-values for multiplicity in order to control the false discovery
rate (FDR, default method);

.. code-block:: sh

  mictools adjust $ODIR/pval.txt $ODIR

The command returns in the OUTPUT directory the following files: 

``pval_adj.txt`` 
  adjusted p values for each variable pair tested, in the same format of 
  ``pval.txt``

``pi0_None.png``
  since the correction method is the Storey's qvalue, the command returns
  a plot with the estimated pi_0 versus the tuning parameter lambda:

  .. image:: docs/images/pi0_None.png

Strength of significant associations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Finally, the strengths of the relationships called significant are estimated 
using MIC_e. By default the significance level is set to 0.05:

.. code-block:: sh

  mictools strength $X $ODIR/pval_adj.txt $ODIR/strength.txt


The output file ``strength.txt`` is a TAB-delimited file, containing for each 
significant association the (corrected) TIC_e p-values, the Pearson's
correlations, the Spearman's coefficients and finally the strengths, *i.e.* the
MIC_e values:

===== ========== ========== ============ ========= =========== ========
Class Var1       Var2       TICePVal     PearsonR  SpearmanRho MICe
===== ========== ========== ============ ========= =========== ========
None  bullseye_x bullseye_y 3.833704e-02 -0.068586 -0.078734   0.424553
None  circle_x   circle_y   4.723013e-04 -0.068343 -0.077292   0.631458
None  dots_x     dots_y     1.983666e-02 -0.060342 -0.126174   0.500185
None  slant_up_x slant_up_y 1.593666e-02 -0.068609 -0.086098   0.355019
None  star_x     star_y     4.723013e-04 -0.062961 -0.051445   0.633117
None  x_shape_x  x_shape_y  4.723013e-04 -0.065583 -0.020535   0.566703
===== ========== ========== ============ ========= =========== ========



.. [Reshef2016] Yakir A. Reshef, David N. Reshef, Hilary K. Finucane and 
                Pardis C. Sabeti and Michael Mitzenmacher. Measuring Dependence
                Powerfully and Equitably. Journal of Machine Learning Research, 
                2016.
.. [Matejka2017] J. Matejka and G. Fitzmaurice. Same Stats, Different Graphs: 
                 Generating Datasets with Varied Appearance and 
                 Identical Statistics through Simulated Annealing. 
                 ACM SIGCHI Conference on Human Factors in Computing Systems, 2017.


CHANGES.rst
CHANGES
=======

1.1.0
-----
Python 2.7 support ending (Python 3.5+ supported).

1.0.2
-----
Fix pandas DataFrame as_matrix() warning/error.

1.0.1
-----
Fix documentation for PYPI.

1.0.0
-----
First stable release.

mictools/null.py
##    Copyright 2017 MICtools Developers <davide.albanese@gmail.com>

##    This file is part of MICtools.
##
##    MICtools is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    MICtools is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.

##    You should have received a copy of the GNU General Public License
##    along with MICtools. If not, see <http://www.gnu.org/licenses/>.


import sys

import numpy as np
import pandas as pd

import minepy
import mictools.utils
from mictools import NULL_HIST_RES


def compute_null_oneclass(X, Y=None, rowwise=False, B=9, c=5, nperm=250000, seed=0):

    mictools.utils.check_data(X, Y=Y)

    bins = np.linspace(0, 1, NULL_HIST_RES+1)
    hist = np.zeros(NULL_HIST_RES, dtype=np.int64)
    rs = np.random.RandomState(seed)
    mine = minepy.MINE(alpha=B, c=c, est="mic_e")

    Xa = X.values
    if Y is None:
        idx = np.arange(Xa.shape[0])
        Ya, max_idx_rowwise = None, None
    else:
        Ya = Y.values
        if rowwise:
            max_idx_rowwise = min(Xa.shape[0], Ya.shape[0])
            idx = None

    for n in range(nperm):
        if Y is None:
            i, j = rs.choice(idx, size=2, replace=False)
            x, y = Xa[i], rs.permutation(Xa[j])     
        else:
            if rowwise:
                i = j = rs.randint(max_idx_rowwise)
            else:
                i, j = rs.randint(Xa.shape[0]), rs.randint(Ya.shape[0])

            x, y = Xa[i], rs.permutation(Ya[j])

        mine.compute_score(x, y)
        tic = mine.tic(norm=True)
        hist_idx = min(np.digitize([tic], bins)[0]-1, NULL_HIST_RES-1)
        hist[hist_idx] += 1

    # right-tailed area
    hist_cum = np.cumsum(hist[::-1])[::-1]

    index = pd.MultiIndex.from_arrays([bins[:-1], bins[1:]],
                                      names=('BinStart', 'BinEnd'))
    hist_df = pd.DataFrame({"NullCount": hist, 
                            "NullCountCum": hist_cum},
                            index=index,
                            columns=["NullCount", "NullCountCum"])

    return hist_df


def compute_null(X, labels=None, Y=None, rowwise=False, B=9, c=5, nperm=250000,
                 seed=0):

    mictools.utils.check_data(X, labels=labels, Y=Y)

    if labels is None:
        labels = pd.Series('None', index=X.columns)

    clss = sorted(labels.unique())
    null_dist_list = []
    for cl in clss:
        keep = (cl == labels)
        X_cl = X.loc[:, keep]
        
        if Y is None:
            Y_cl = None
        else:
            Y_cl = Y.loc[:, keep]
            
        null_dist_cl = compute_null_oneclass(
            X_cl, Y_cl, rowwise=rowwise, B=B, c=c, nperm=nperm, seed=seed)

        null_dist_list.append(null_dist_cl)

    null_dist = pd.concat(null_dist_list, axis=0, keys=clss, names=["Class"])

    return null_dist


def null_cmd(xvars_fn, output_fn, labels_fn=None, target=None, yvars_fn=None,
             rowwise=False, nperm=250000, seed=0, grid=9, clumps=5):

    X, labels, Y = mictools.utils.read_data(xvars_fn, labels_fn, target,
                                            yvars_fn)    

    sys.stdout.write("X: {} variables\n".format(X.shape[0]))

    if not Y is None:
        sys.stdout.write("Y: {} variables\n".format(Y.shape[0]))
    
    sys.stdout.write("Samples per class:\n")
    if not labels is None:
        for cl in sorted(labels.unique()):
            sys.stdout.write("* {}: {:d}\n".format(cl, (cl == labels).sum()))
    else:
        sys.stdout.write("* None: {:d}\n".format(X.shape[1]))

    null_dist = compute_null(X=X,
                             labels=labels, 
                             Y=Y, 
                             rowwise=rowwise,
                             B=grid, 
                             c=clumps,
                             nperm=nperm,
                             seed=seed)
    
    null_dist.to_csv(output_fn, sep='\t', float_format="%.6f")


def mergenull_cmd(null_fns, output_fn):
    null_dist = pd.read_csv(null_fns[0], sep='\t', index_col=[0, 1, 2])
    
    for null_fn in null_fns[1:]:
        null_dist_tmp = pd.read_csv(null_fn, sep='\t', index_col=[0, 1, 2])
        null_dist += null_dist_tmp

    null_dist.to_csv(output_fn, sep='\t')


mictools/pval.py
##    Copyright 2017 MICtools Developers <davide.albanese@gmail.com>

##    This file is part of MICtools.
##
##    MICtools is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    MICtools is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.

##    You should have received a copy of the GNU General Public License
##    along with MICtools. If not, see <http://www.gnu.org/licenses/>.


import os
import os.path
import sys
import itertools

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

import minepy
import mictools.utils
import mictools.mtest
from mictools import NULL_HIST_RES


def compute_pval_oneclass(X, null_dist, Y=None, single=False, B=9, c=5):

    mictools.utils.check_data(X, Y=Y)

    bins = np.linspace(0, 1, NULL_HIST_RES+1)

    # observed values/distribution
    names=['Var1', 'Var2']
    Xa = X.values
    if Y is None:
        _, tic = minepy.pstats(Xa, alpha=B, c=c, est="mic_e")
        index = pd.MultiIndex.from_tuples(
            list(itertools.combinations(X.index, 2)), names=names)
    else:
        Ya = Y.values
        if single:
            _, tic = mictools.utils.sstats(Xa, Ya, alpha=B, c=c, est="mic_e")
            index = pd.MultiIndex.from_arrays([X.index, Y.index], names=names)
        else:
            _, tic = minepy.cstats(Xa, Ya, alpha=B, c=c, est="mic_e")
            index = pd.MultiIndex.from_product([X.index, Y.index], names=names)
            tic = tic.flatten()

    observed_hist = np.histogram(tic, bins)[0].astype(np.int64)

    # right-tailed area
    observed_hist_cum = np.cumsum(observed_hist[::-1])[::-1]

    # p-values
    null_hist_cum = null_dist["NullCountCum"].values
    pval = (np.interp(tic, bins[:-1], null_hist_cum) + 1) / \
        (null_hist_cum[0] + 1)
    pval = pd.Series(pval, index=index)
    
    # observed
    obs = pd.Series(tic, index=index)
    
    # distribution
    index = pd.MultiIndex.from_arrays([bins[:-1], bins[1:]],
                                      names=('BinStart', 'BinEnd'))
    obs_dist = pd.DataFrame({"ObsCount": observed_hist,
                             "ObsCountCum": observed_hist_cum},
                            index=index,
                            columns=["ObsCount", "ObsCountCum"])

    return obs_dist, obs, pval


def compute_pval(X, null_dist, labels=None, Y=None, single=False, B=9, c=5):

    mictools.utils.check_data(X, labels=labels, Y=Y)

    if labels is None:
        labels = pd.Series('None', index=X.columns)

    clss = sorted(labels.unique())
    obs_dist_list, obs_list, pval_list = [], [], []
    for cl in clss:
        keep = (cl == labels)
        X_cl = X.loc[:, keep]
        null_dist_cl = null_dist.loc[cl]

        if Y is None:
            Y_cl = None
        else:
            Y_cl = Y.loc[:, keep]
            
        obs_dist_cl, obs_cl, pval_cl = compute_pval_oneclass(
            X_cl, null_dist_cl, Y_cl, single=single, B=B, c=c)

        obs_cl.name = cl
        pval_cl.name = cl
        obs_dist_list.append(obs_dist_cl)
        obs_list.append(obs_cl)
        pval_list.append(pval_cl)
    
    obs_dist = pd.concat(obs_dist_list, axis=0, keys=clss, names=["Class"])
    obs = pd.concat(obs_list, axis=1)
    pval = pd.concat(pval_list, axis=1)
    
    return obs_dist, obs, pval


def plot_pval(pval, output_dir):

    BINS = 50

    for cl in pval.columns:
        fig = plt.figure(figsize=(10, 4))
        ax1 = plt.subplot(111)
        bins = np.linspace(0, 1, BINS+1)
        hist, bin_edges = np.histogram(pval[cl], bins=bins, density=True)
        plt.bar(bin_edges[:-1], hist, width=1/BINS, log=False,
                linewidth=1, color="white", edgecolor="black")
        plt.xlabel("p-value")
        plt.ylabel("Density")

        output_fn = os.path.join(output_dir, "pval_{}.png".format(cl))
        fig.savefig(output_fn, bbox_inches='tight', dpi=300, format='png')


def pval_cmd(xvars_fn, null_fn, output_dir, labels_fn=None, target=None,
             yvars_fn=None, single=False, grid=9, clumps=5):

    X, labels, Y = mictools.utils.read_data(xvars_fn, labels_fn, target,
                                            yvars_fn)    
    null_dist = pd.read_csv(null_fn, sep='\t', index_col=[0, 1, 2])

    obs_dist, obs, pval = compute_pval(X=X,
                                       null_dist=null_dist,
                                       labels=labels, 
                                       Y=Y, 
                                       single=single,
                                       B=grid, 
                                       c=clumps)

    ntotperm = null_dist["NullCountCum"].iloc[0]
    minpval = 1 / ntotperm
    sys.stdout.write("The minimum p-value with a total of {:d} permutations is "
                     "{:e}\n".format(ntotperm, minpval))

    try:
        os.makedirs(output_dir)
    except OSError:
        if not os.path.isdir(output_dir):
            sys.stderr.write("directory {} cannot be created\n".\
                             format(output_dir))
            exit(1)
    
    obs_dist.to_csv(os.path.join(output_dir, "obs_dist.txt"), sep='\t',
                    float_format="%.6f")
    obs.to_csv(os.path.join(output_dir, "obs.txt"), sep='\t',
               float_format='%.6f')
    pval.to_csv(os.path.join(output_dir, "pval.txt"), sep='\t',
                float_format='%e')
    plot_pval(pval, output_dir)


def plot_pi0(pi0, pi0_lmb, lmb, pi0_smooth, output_fn):
    fig = plt.figure(figsize=(6, 6))
    ax1 = plt.subplot(111)
    plt.plot(lmb, pi0_lmb, 'ko')
    if not pi0_smooth is None:
        plt.plot(lmb, pi0_smooth, 'k', label='Quadratic smoothing')
    plt.axhline(pi0, linestyle='--', color='r', label='pi0')
    plt.xlabel("lambda")
    plt.ylabel("pi0(lambda)")
    plt.legend()
    fig.savefig(output_fn, bbox_inches='tight', dpi=300, format='png')


def adjust_cmd(pval_fn, output_dir, method='qvalue'):

    pval = pd.read_csv(pval_fn, sep='\t', index_col=[0, 1])
    pval_adj = pd.DataFrame(index=pval.index, columns=pval.columns)
    
    try:
        os.makedirs(output_dir)
    except OSError:
        if not os.path.isdir(output_dir):
            sys.stderr.write("directory {} cannot be created\n".\
                             format(output_dir))
            exit(1)

    for cl in pval.columns:
        if method == 'qvalue':
            pval_adj[cl], pi0, pi0_lmb, lmb, pi0_smooth = \
                mictools.mtest.qvalue(pval[cl])
            output_plot_fn = os.path.join(output_dir, "pi0_{}.png".format(cl))
            plot_pi0(pi0, pi0_lmb, lmb, pi0_smooth, output_plot_fn)
        else:
            pval_adj[cl] = mictools.mtest.multipletests(pval[cl], method=method)
    
    output_fn = os.path.join(output_dir, "pval_adj.txt")
    pval_adj.to_csv(output_fn, sep='\t', float_format='%e')

mictools/utils.py
##    Copyright 2017 MICtools Developers <davide.albanese@gmail.com>

##    This file is part of MICtools.
##
##    MICtools is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    MICtools is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.

##    You should have received a copy of the GNU General Public License
##    along with MICtools. If not, see <http://www.gnu.org/licenses/>.


import numpy as np
import pandas as pd

import minepy


def sstats(X, Y, alpha=0.6, c=15, est="mic_approx"):
    mic, tic = [], []
    mine = minepy.MINE(alpha=alpha, c=c, est=est)
    for i in range(min(X.shape[0], Y.shape[0])):
        mine.compute_score(X[i], Y[i])
        mic.append(mine.mic())
        tic.append(mine.tic(norm=True))
    mic, tic = np.asarray(mic), np.asarray(tic)

    return mic, tic


def read_table(input_fn):
    """Read table.
    """
    
    table = pd.read_csv(input_fn, sep='\t', index_col=0)
    # cast index into string
    table.index = [str(elem) for elem in table.index]

    return table


def check_data(X, Y=None, labels=None):
    if not (Y is None):
        if X.shape[1] != Y.shape[1]:
            raise ValueError("different number of samples between X and Y")

        if not (X.columns == Y.columns).all():
            raise ValueError("sample names mismatch between X and Y")

    if not (labels is None):
        if X.shape[1] != labels.shape[0]:
            raise ValueError("different number of samples between X and labels")

        if not (X.columns == labels.index).all():
            raise ValueError("sample names mismatch between X and labels")


def read_data(xvars_fn, labels_fn=None, target=None, yvars_fn=None):

    X = read_table(xvars_fn)
    sample_ids = sorted(X.columns)
    labels = None
    Y = None
    
    if not labels_fn is None:
        if target is None:
            raise ValueError("target (labels file column) is required when the "
                             "labels file is provided")

        labels_df = read_table(labels_fn)

        if not (target in labels_df.columns):
            raise Exception("target {} is not in the labels file".\
                            format(target))
        
        labels = labels_df.loc[labels_df[target].notnull(), target]
        sample_ids = sorted(set(sample_ids) & set(labels.index))

    if not yvars_fn is None:
        Y = read_table(yvars_fn)
        sample_ids = sorted(set(sample_ids) & set(Y.columns))
        Y = Y[sample_ids]
    
    X = X[sample_ids]

    if not labels_fn is None:
        labels = labels[sample_ids]
        
    return X, labels, Y

scripts/mictools_cmd.py
#! /usr/bin/env python

##    Copyright 2016 Davide Albanese <davide.albanese@gmail.com>
##    Copyright 2016 Fondazione Edmund Mach (FEM)

##    This file is part of strainest.
##
##    strainest is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    strainest is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.

##    You should have received a copy of the GNU General Public License
##    along with strainest.  If not, see <http://www.gnu.org/licenses/>.

import os
import os.path

import click

import mictools.null
import mictools.pval
import mictools.strength
from mictools import __version__


labels_help = 'tab-delimited sample labels file (samples x metadata). If this ' \
              'file is provided, the analysis will be performed within each ' \
              'class independently. In order to obtain meaningful results, the ' \
              'number of samples in each class should be roughly the same. The ' \
              'first column must contain the sample IDs.'

target_help = 'labels file column to be used in the per-class analysis (see ' \
              '-l/--labels).'

yvars_help = 'tab-delimited data file (K x N). If this file is provided, the ' \
             'analysis will be performed between all variables of the dataset ' \
             'XVARS and of those of the dataset YVARS.'

rowwise_help = 'when the dataset Y is provided, compute the TICe values between ' \
               'XVARS_i and YVARS_i for each i, instead of XVARS_i and YVARS_j ' \
               'for each i, j.'
    

@click.group()
def cli():
    """MICtools v. {}.""".format(__version__)
    pass

@cli.command()
@click.argument('xvars', type=click.Path(exists=True))
@click.argument('output', type=click.Path(exists=False, writable=True))
@click.option('-l', '--labels', type=click.Path(exists=True), help=labels_help)
@click.option('-t', '--target', type=click.STRING, help=target_help)
@click.option('-y', '--yvars', type=click.Path(exists=True), help=yvars_help)
@click.option('-r', '--rowwise', is_flag=True, help=rowwise_help)
@click.option('-p', '--nperm', type=click.INT, default=200000,
              show_default=True, help='number of permutations.')
@click.option('-s', '--seed', type=click.INT, default=0,
              show_default=True, help="random seed.")
@click.option('-b', '--grid', type=click.INT, default=9,
              show_default=True, help='TICe maximum grid size B(n). Set to 4 '
              'or 6 for less complex alternative hypotheses and to 12 or more '
              'for more complex alternative hypotheses.')
@click.option('-c', '--clumps', type=click.INT, default=5,
              show_default=True, help="TICe c parameter.")
def null(xvars, output, labels, target, yvars, rowwise, nperm, seed, grid,
         clumps):
    """Compute the TICe null distribution.

    XVARS is a tab-delimited data file. The file must contain the variables by
    rows and the samples by columns (M x N columns). The first column
    must contain the variable IDs and the first row the sample IDs. By default,
    all the possible variable pairs ((M*(M-1))/2) are tested.
    
    If YVARS (K x N) is provided, the analysis will be performed between all 
    variables of the dataset XVARS and of those of the dataset YVARS, for a
    total of M x K variable pairs. When the option -r/--rowwise is set, only the
    pairs XVARS_i and YVARS_i for each i are tested.

    If sample classes are provided (see -l/--labels and -t/--target), the
    analysis will be performed within each class independently. Otherwise, a
    single class will be named None.   

    The command returns the TICe null distribution (OUTPUT) for each class.

    Example:

        mictools null data1.txt null_dist.txt -y data2.txt

    will compute the empirical null distribution considering all the variable
    pairs between the dataset data1.txt and the dataset data2.txt.
    """

    mictools.null.null_cmd(xvars, output, labels, target, yvars, rowwise, nperm,
                           seed, grid, clumps)


@cli.command()
@click.argument('null', nargs=-1, type=click.Path(exists=True))
@click.argument('output', nargs=1, type=click.Path(exists=False, writable=True))
def mergenull(null, output):
    """Merge multiple TICe null distributions.

    NULL is the input null distribution(s) and output is the output null
    distribution.

    Example:

        mictools mergenull null_dist_1.txt null_dist_2.txt null_dist.txt

    will merge the independent null distributions null_dist_1.txt and 
    null_dist_2.txt into the output file null_dist.txt.
    """

    mictools.mergenull.mergenull_cmd(null, output)


@cli.command()
@click.argument('xvars', type=click.Path(exists=True))
@click.argument('null', type=click.Path(exists=True))
@click.argument('output', type=click.Path(writable=True, dir_okay=True,
                resolve_path=True))
@click.option('-l', '--labels', type=click.Path(exists=True), help=labels_help)
@click.option('-t', '--target', type=click.STRING, help=target_help)
@click.option('-y', '--yvars', type=click.Path(exists=True), help=yvars_help)
@click.option('-r', '--rowwise', is_flag=True, help=rowwise_help)
@click.option('-b', '--grid', type=click.INT, default=9,
              show_default=True, help='TICe maximum grid size B(n). Set to 4 '
              'or 6 for less complex alternative hypotheses and to 12 or more '
              'for more complex alternative hypotheses.')
@click.option('-c', '--clumps', type=click.INT, default=5,
              show_default=True, help="TICe c parameter.")

def pval(xvars, null, output, labels, target, yvars, rowwise, grid, clumps):
    """Compute TICe p-values.

    XVARS is a tab-delimited data file. The file must contain the variables by
    rows and the samples by columns (M x N columns). The first column
    must contain the variable IDs and the first row the sample IDs. By default,
    all the possible variable pairs ((M*(M-1))/2) are tested.

    NULL is the empirical TICe null distribution (see the null subcommand).
    
    If YVARS (K x N) is provided, the analysis will be performed between all 
    variables of the dataset XVARS and of those of the dataset YVARS, for a
    total of M x K variable pairs. When the option -r/--rowwise is set, only the
    pairs XVARS_i and YVARS_i for each i are tested.

    If sample classes are provided (see -l/--labels and -t/--target), the
    analysis will be performed within each class independently. Otherwise, a
    single class will be named None.

    The command returns in the OUTPUT directory the following files:
        
    \b
    - obs_dist.txt: the observed TICe distribution for each class;
    - obs.txt: the observed TICe values for each variable pair tested in 
               each class;
    - pval.txt: empirical p values for each variable pair tested in 
                each class;
    - pval_CLASS.png: the p values distribution for each class.

    Example:

        mictools pval data1.txt null_dist.txt pval.txt . -y data2.txt

    will compute the TICe p values for each variable pair between the dataset
    data1.txt and the dataset data2.txt.
    """
    
    mictools.pval.pval_cmd(xvars, null, output, labels, target, yvars, rowwise, 
                           grid, clumps)


@cli.command()
@click.argument('pval', type=click.Path(exists=True))
@click.argument('output', type=click.Path(writable=True, dir_okay=True,
                resolve_path=True))
@click.option('-m', '--method', default='qvalue', help='correction method.')
def adjust(pval, output, method):
    """Multiple testing correction.

    PVAL is the input file containing the uncorrected p values (see the 
    pval subcommand). The command returns in the OUTPUT directory the following 
    files: 

    \b
    - pval_adj.txt: adjusted p values for each variable pair tested in 
                    each class.
    - pi0_CLASS.png: if the method (-m/--method) is the qvalue, command will
                     return the estimated pi_0 versus the tuning parameter
                     lambda for each class. 

    The available methods are:

    \b
    - bonferroni: one-step correction
    - sidak: one-step correction
    - holm-sidak: step down method using Sidak adjustments
    - holm: step-down method using Bonferroni adjustments
    - simes-hochberg: step-up method  (independent)
    - hommel: closed method based on Simes tests (non-negative)
    - fdr_bh: Benjamini/Hochberg  (non-negative)
    - fdr_by: Benjamini/Yekutieli (negative)
    - fdr_tsbh: two stage fdr correction (non-negative)
    - fdr_tsbky: two stage fdr correction (non-negative)
    - qvalue: Storey's q value (smoother method)

    Example:

        mictools adjust pval.txt .
    """

    mictools.pval.adjust_cmd(pval, output, method)


@cli.command()
@click.argument('xvars', type=click.Path(exists=True))
@click.argument('pval', type=click.Path(exists=True))
@click.argument('output', type=click.Path(exists=False, writable=True))
@click.option('-l', '--labels', type=click.Path(exists=True), help=labels_help)
@click.option('-t', '--target', type=click.STRING, help=target_help)
@click.option('-y', '--yvars', type=click.Path(exists=True), help=yvars_help)
@click.option('-p', '--thr', type=click.FLOAT, default=0.05,
              show_default=True, help="significance threshold.")
@click.option('-a', '--alpha', type=click.FLOAT, default=None,
              show_default=True, help="MICe alpha parameter. Automatically "
              "chosen by default.")
@click.option('-c', '--clumps', type=click.INT, default=5,
              show_default=True, help="MICe c parameter.")
def strength(xvars, pval, output, labels, target, yvars, thr, alpha, clumps):
    """Compute the strength (MICe).
    
    Compute the strength (MICe) on the variable pairs called significant at
    the significance threshold THR (parameter -p/--thr).
    
    XVARS is a tab-delimited data file. The file must contain the variables by
    rows and the samples by columns (M x N columns). The first column
    must contain the variable IDs and the first row the sample IDs. By default,
    all the possible variable pairs ((M*(M-1))/2) are tested.
    
    If YVARS (K x N) is provided, the analysis will be performed between all 
    variables of the dataset XVARS and of those of the dataset YVARS, for a
    total of M x K variable pairs.

    If sample classes are provided (see -l/--labels and -t/--target), the
    analysis will be performed within each class independently. In this case,
    the relationships called significant in at least one class are evaluated.

    The OUTPUT file contains MICe, the Pearson and the Spearman's rank 
    correlation coefficients and the TICe p values in each class.

    Example:

        mictools strength data1.txt pval_adj.txt strength.txt -y data2.txt
    """

    mictools.strength.strength_cmd(xvars, pval, output, labels, target, yvars,
                                   thr, alpha, clumps)
                                   